src/once_cell.rs

```diff
@@ -0,0 +1,800 @@
+use std::cell::UnsafeCell;
+use std::convert::Infallible;
+use std::fmt;
+use std::future::Future;
+use std::mem::{forget, MaybeUninit};
+use std::pin::Pin;
+use std::ptr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+
+use event_listener::{Event, EventListener};
+use futures_lite::future;
+
+/// The `OnceCell` is uninitialized.
+const UNINITIALIZED: usize = 0;
+/// The `OnceCell` is being initialized.
+const INITIALIZING: usize = 1;
+/// The `OnceCell` is initialized.
+const INITIALIZED: usize = 2;
+
+/// A memory location that can be written to at most once.
+///
+/// A `OnceCell` can be used to store a single value, and only once. However,
+/// once the value is stored, it can be accessed directly through a reference
+/// instead of needing an RAII guard like `Mutex` or `RwLock`.
+///
+/// # Examples
+///
+/// This structure is useful for a variety of patterns, most notably for one-time
+/// initialization.
+///
+/// ```rust
+/// use async_lock::OnceCell;
+///
+/// # struct Foobar;
+///
+/// async fn very_expensive_initialization() -> Foobar {
+///     // Imagine this is very expensive to initialize,
+///     // for instance, it requires a network request or
+///     // a database call.
+///     # Foobar
+/// }
+///
+/// struct LazyFoobar {
+///     inner: OnceCell<Foobar>,
+/// }
+///
+/// impl LazyFoobar {
+///     fn new() -> Self {
+///         Self {
+///             inner: OnceCell::new(),
+///         }
+///     }
+///
+///     async fn load(&self) -> &Foobar {
+///         self.inner.get_or_init(|| async {
+///             very_expensive_initialization().await
+///         }).await
+///     }
+/// }
+/// ```
+pub struct OnceCell<T> {
```

Comment:
It is interesting that the standard library has renamed this to OnceLock.
https://doc.rust-lang.org/nightly/std/sync/struct.OnceLock.html

Reply:
Interesting, it looks like the name is very much in flux for `std` right now. Do we wait for libstd to arrive at a proper conclusion, or do we just keep `OnceCell`?

Reply:
OnceCell seems good to me. Other crates use the same name as well.

Reply:
@taiki-e Any thoughts on this? Right now this requested change is blocking the merge.

Reply:
Future notes: `OnceLock` was decided on the final name by `libstd`. It might be prudent to rename `OnceCell` appropriately at the next breaking change.